discord/asset.py

```diff
@@ -0,0 +1,24 @@
+import io
+from typing import Union, TYPE_CHECKING
+
+from pydantic import BaseModel
+
+
+if TYPE_CHECKING:
+    from discord import Client
+
+
+class Asset(BaseModel):
+    _client: "Client"
+    url: str
+
+    async def read(self):
+        return await self._client.httphandler.get_from_cdn(self.url)
+
+    async def save(self, fp: Union[str, io.BufferedIOBase]):
+        data = await self.read()
+        if isinstance(fp, str):
```

Comment:
I would first check if fp is a valid write path to reduce network activity.

Reply:
What method do you suggest to use to check if the given path is valid?

Reply:
First try to do the open() call, and when that works do the self.read() call. If the file is invalid, open() will throw an error. Just let it propagate as it should be clearly stating the file does not exist to the user.